### DISCUSSION TITLE
LSM6DSL Sample Fails to Reboot LSM6DS3TR-C (Pin-for-pin compatible with LSM6DSL) on XIAO BLE Sense

### DISCUSSION THREAD

**Discussion by MattMacGregor** [zephyrproject-rtos/zephyr#56042] [Q&A]
New to Zephyr so apologies if there are any violations of contribution rules/standards in this issue.

I'm trying to get the IMU on the XIAO BLE Sense up and running with the [lsm6dsl sample](drivers/sensor/lsm6dsl/lsm6dsl.c) but it seems to fail during the initialization section of the [LSM6DSL](https://www.st.com/content/st_com/en/products/mems-and-sensors/inemo-inertial-modules/lsm6dsm.html) driver. As the [LSM6DS3TR-C](https://www.st.com/en/mems-and-sensors/lsm6ds3tr-c.html) is pretty much a pin-for-pin replacement, I believe this should be working, so something seems amiss (but I acknowledge this sample hasn't been officially tested on the XIAO BLE Sense yet).

It seems to fail at these lines https://github.com/zephyrproject-rtos/zephyr/blob/b0ea0c8ec6135aa5cb8d0e70ca72af2abe1f38fa/drivers/i2c/i2c_nrfx_twim.c#L155-L178 in the `nrfx_twim` driver after calling lsm6dsl_reboot here: https://github.com/zephyrproject-rtos/zephyr/blob/9a63f9d3d8fe9d1c2e2e6e38b4d74dfb07f54abc/drivers/sensor/lsm6dsl/lsm6dsl.c#L716

I2C debug log messages show I2C messages repeatedly being dropped when running the sample, and since the reboot pin process seems to be the same between the two chips, my gut is pointing me towards the I2C being the issue.

I added a log message to the event_handler for the `nrfx_twim` and didn't see any messages from it. Add that it fails to take the `completion_sync` mutex, which is released by the `event_handler` on `NRFX_TWIM_EVT_DONE`, seems to indicate that the `event_handler` is not being successfully called. FYI, `k_sem_take` always fails with `-EAGAIN`, but I don't believe that is much help in this case. Maybe it's my lack of experience but I cannot seem to find why this would be the case.

To reproduce simply build the `lsm6dsl` sample ([samples/sensor/lsm6dsl](https://github.com/zephyrproject-rtos/zephyr/tree/b0ea0c8ec6135aa5cb8d0e70ca72af2abe1f38fa/samples/sensor/lsm6dsl)) for the board `xiao_ble_sense`. It is using all the default configuration settings.

**Comment by MattMacGregor** [comment]
This was rightfully converted to a discussion, as the problem seems to be more focused in the I2C realm then this sensor. Here is a link to the issue with a bit more detail: https://github.com/zephyrproject-rtos/zephyr/issues/56012

**Reply by tarikgraba** [reply]
@MattMacGregor  with PR #56640 applied, I can access the xiao sens imu using i2c commands from a Zephyr shell (starting with samples/basic/minimal), still the the lsm6dsl sample does not work. 
Any hints?

**Reply by tarikgraba** [reply]
I have seen from the PR discussion that the regulator boot delay has also to be changed (from 900 to 3000) which makes I2C communication and the example work correctly.

**Reply by tarikgraba** [reply]
PR #57875 changes the default boot delay of the regulator.

**Comment by purpl3F0x** [comment]
Hello, is this supposed to be working now ? Cause I still get `[00:00:00.500,762] <err> i2c_nrfx_twim: Error on I2C line occurred for message 0
[00:00:00.500,915] <err> LSM6DSL: Failed to initialize chip`, using ncs 2.4.0 with a modified xiao ble sense dts

**Reply by syamusivan** [reply]
Increase "startup-delay-us" to some higher value (10000 us worked for me) within xiao_ble_sense.dts. The delay was around 900 us initially. I have not tested the optimal delay (higher than 900 us).

**Reply by purpl3F0x** [reply]
Wasn't the delay. I was falsly enabling pull-up.

**Reply by syamusivan** [reply]
Could you please send me the code snippet that was falsly enabling pull-up. This issue is reappearing for me.
Thank you

**Reply by mentarus** [reply]
Still seems to be an issue, posted about it on the Nordic forum [here](https://devzone.nordicsemi.com/f/nordic-q-a/109732/running-the-lsm6dls-imu-zephyr-example-with-nrf52840-based-xiao-ble-sense/476976). Note that the Arduino libraries work fine, I confirmed that this is not a hardware issue as I can access the IMU fine through the [SEEED provided](https://github.com/Seeed-Studio/Seeed_Arduino_LSM6DS3) library.
